### PR TITLE
Add `nerdctl run --env-file <FILE>`

### PR DESCRIPTION
**Pull request by ilyee**
Signed-off-by: Zuti He <ilyeeelihe@gmail.com>

Related to #215

### CODE REVIEW

File: run.go [hunk 1]
```diff
@@ -783,3 +796,26 @@ func writeCIDFile(path, id string) error {
...
+	vars := make([]string, 0)
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, errors.Errorf("failed to open env file %s: %v", path, err)
+	}
+	defer f.Close()
+
+	sc := bufio.NewScanner(f)
+	for sc.Scan() {
+		line := sc.Text()
```
Comment (AkihiroSuda) [inline run.go:810]: strings.TrimSpace(line)

File: run.go [hunk 2]
```diff
@@ -783,3 +796,26 @@ func writeCIDFile(path, id string) error {
 		return err
 	}
 }
+
+func getEnvVars(path string) ([]string, error) {
```
Comment (AkihiroSuda) [inline run.go:800]: Func name should be like "parseEnvFile"

File: run_test.go [hunk 3]
```diff
@@ -157,3 +157,24 @@ func TestRunShmSize(t *testing.T) {
 
 	base.Cmd("run", "--rm", "--shm-size", shmSize, testutil.AlpineImage, "/bin/grep", "shm", "/proc/self/mounts").AssertOutContains("size=32768k")
 }
+
+func TestRunEnvFile(t *testing.T) {
+	base := testutil.NewBase(t)
+	const fileName = "env.file"
```
Comment (AkihiroSuda) [inline run_test.go:163]: https://golang.org/pkg/io/ioutil/#TempFile

File: run_test.go [hunk 4]
```diff
@@ -157,3 +157,24 @@ func TestRunShmSize(t *testing.T) {
...
+
+	f, err := os.Create(fileName)
```
Comment (AkihiroSuda) [inline run_test.go:165]: https://golang.org/pkg/io/ioutil/#WriteFile

File: run.go [hunk 5]
```diff
@@ -238,6 +239,10 @@ var runCommand = &cli.Command{
 			Name:  "shm-size",
 			Usage: "Size of /dev/shm",
 		},
+		&cli.StringFlag{
```
Comment (AkihiroSuda) [inline run.go:242]: Perhaps StringSliceFlag to support multiple files

Comment (AkihiroSuda) [reply run.go:242]: Also, should be defined right after "env" flag

File: run.go [hunk 6]
```diff
@@ -783,3 +796,26 @@ func writeCIDFile(path, id string) error {
 		return err
 	}
 }
+
+func getEnvVars(path string) ([]string, error) {
+	vars := make([]string, 0)
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, errors.Errorf("failed to open env file %s: %v", path, err)
```
Comment (AkihiroSuda) [inline run.go:804]: errors.Wrapf

File: run.go (hunk 5 above)
Comment (ilyee) [reply run.go:242]: > Perhaps StringSliceFlag to support multiple files

Agree with that

File: run_test.go [hunk 7]
```diff
@@ -157,3 +157,29 @@ func TestRunShmSize(t *testing.T) {
...
+	file2, err := ioutil.TempFile("", pattern)
+	assert.NilError(base.T, err)
+	path2 := file2.Name()
+	defer file2.Close()
+	defer os.Remove(path2)
+	err = ioutil.WriteFile(path2, []byte("# this is a comment line\nTESTKEY2=TESTVAL2"), 0666)
+	assert.NilError(base.T, err)
+
+	base.Cmd("run", "--rm", "--env-file", path1, "--env-file", path2, testutil.AlpineImage,
+		"/usr/bin/env", "|", "grep", "TESTKEY1", "|", "awk", "-F=", "'{print $2}'").AssertOutContains("TESTVAL1")
```
Comment (AkihiroSuda) [inline run_test.go:182]: Why not just echo $TESTKEY1

File: run.go [hunk 8]
```diff
@@ -215,6 +216,10 @@ var runCommand = &cli.Command{
 			Aliases: []string{"e"},
 			Usage:   "Set environment variables",
 		},
+		&cli.StringSliceFlag{
+			Name:  "env-file",
+			Usage: "Set environment variables from file",
```
Comment (fahedouch) [inline run.go:221]: files(s)

File: run_test.go (hunk 7 above)
Comment (ilyee) [reply run_test.go:182]: Sry for the mistake I made, but here I found when I use echo as init command, the output env var is not correct, same thing happens using `env` flag like:
`nerdctl --namespace=nerdctl-test run --rm -e TESTKEY=TESTVALUE ghcr.io/stargz-containers/alpine:3.13-org echo $TESTKEY`
Output should be `TESTVALUE` but actually nothing. And when I attached to the container and run `echo $TESTKEY`, it seems working. (this also happens with `docker run`). And then I found host env will be outputed when echo is init command by `echo $PATH`.

Is it normal? Or something wrong with my host's env or config?

Comment (AkihiroSuda) [reply run_test.go:182]: Probably you forgot to quote `'$TESTKEY'`

Comment (ilyee) [reply run_test.go:182]: > Probably you forgot to quote `'$TESTKEY'`

quote doesn't work, instead `$TESTKEY` will be the output

Comment (AkihiroSuda) [reply run_test.go:182]: "sh" , "-c", "echo '$TESTKEY'"

Comment (fahedouch) [reply run_test.go]: you should escape  the $ with `\` in the runner shell .
Try : 
`nerdctl --namespace=nerdctl-test run --rm -e TESTKEY=TESTVALUE ghcr.io/stargz-containers/alpine:3.13-org sh -c " echo \$TESTKEY"`

Comment (ilyee) [reply run_test.go]: > you should escape the $ with `\` in the runner shell .

thx a lot, that works

Comment (fahedouch) [reply run_test.go]: it looks like test pass without escaping the $.  How did you manage to fix it ?

**Review by AkihiroSuda — APPROVED** [review]

**Comment by AkihiroSuda** [comment]
Thanks

**Pull request merged**